### PR TITLE
Allow event processing cursor to be set in the UI

### PR DESCRIPTION
**Pull request by meyer**
# Problem

1. websqrl only operates on current event streams
2. bluesky demo rules haven't been updated

# Solution

I added a new very much work-in-progress cursor selector UI element to the bottom of the screen. It shows the current selected cursor and shows event processing progress. Check out the attached GIF for the general idea.

I moved event downloading to the web worker. The UI is only responsible for setting the "playhead" aka selected point where event processing should begin. You'll eventually be able to select playback speed to either "realtime" or some arbitrary millisecond value but for now it's processing events as fast as possible (i.e. next tick).

There's a new file, EventDomain, that handles all event-related state in the worker. It's got a few interesting methods:
- `setCursor`: set the date where event processing should begin
- `getNextEventDelay`: get the number of milliseconds to wait before processing the next event
- `processNextEvent`: do the thing

Once all the events for the current cursor have been processed, EventDomain increments the cursor by 10 minutes and then carries on. If the cursor is less than 11 minutes in the future, `getNextEventDelay` returns the number of milliseconds to wait until more data can be fetched.

The path to the date in each event object is now specified via fake jsonpath. I'm not sold on it but it's the only alternative I could think of to letting each demo set its own "getDate" function (can't serialise a function and i don't want to bake demo-specific stuff into the web worker).

none of the names for any of these things are final, feel free to rename things as you please. naming things is hard.

I also added a mock, dev-only Bluesky API so I could test event processing locally without having to wait for 10K events to go by.

# Result

Bluesky demo works now. There are still some bugs, most notably with cursor incrementing when enough time has progressed. I'll be looking into these bugs next.

event processing example

### CODE REVIEW

File: websqrl/pages/bluesky.tsx [hunk 1]
```diff
@@ -15,12 +15,15 @@ export default function TwitterPage() {
         <link rel="icon" href="/favicon.ico" />
       </Head>
       <StreamPage
-        urlPrefix="https://sqrl-twitter-stream.s3.amazonaws.com/v1/"
-        extractDate={(payload) => new Date(payload.dt)}
+        dateJsonPath="timestamp"
+        urlPrefix={
+          process.env.BLUESKY_MOCK_API
+            ? "/api/bluesky-dev/"
```
Comment (qix) [inline websqrl/pages/bluesky.tsx:21]: neat!

File: websqrl/src/EventDomain.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import { getFileNameForDate } from "./getFileNameForDate";
+import type { EventData } from "./types";
+
+// TODO(meyer) actual jsonpath stuff
```
Comment (qix) [inline websqrl/src/EventDomain.ts:4]: I don't think we actually need jsonPath, should be safe to assume the date is in a top level field.

I'd probably just rename this to `getDateFromField` so it's doing what it says, and that's good enough for now?

File: websqrl/src/bluesky-sample.sqrl [hunk 3]
```diff
@@ -14,16 +14,16 @@ LET DatasetId := 1;
...
-LET AuthorCreatedAt := jsonValue(AuthorUserData, "$.created_at");
-LET AuthorName := jsonValue(AuthorUserData, "$.name");
-LET AuthorProfileImageUrl := jsonValue(AuthorUserData, "$.profile_image_url");
+LET Version := jsonValue(EventData, '$.v');
+LET EventName := jsonValue(EventData, '$.eventName');
+LET RecordType := jsonValue(EventData, '$.payload.record.$type');
+LET AuthorId := jsonValue(EventData, '$.payload.author');
+LET TweetText := jsonValue(EventData, '$.payload.record.text');
+LET TweetId := jsonValue(EventData, '$.payload.cid');
+# TODO(meyer) payload.record.createdAt is not available on all events
```
Comment (qix) [inline websqrl/src/bluesky-sample.sqrl:23]: Is this an issue? Fine to leave this in and figure it out later though.

File: websqrl/workers/compile.worker.ts [hunk 4]
```diff
@@ -191,38 +193,64 @@ async function runEvent(event: EventData, requestFeatures: readonly string[]) {
...
+  } else if (message.type === "playhead") {
+    clearTimeout(eventProcessTimer);
+    respond({
+      type: "status",
+      currentIndex: 0,
+      total: 0,
+      timestamp: getFileNameForDate(eventDomain["cursor"]),
+    });
+    eventDomain.setCursor(message.position);
+    queueNextEvent();
```
Comment (qix) [inline websqrl/workers/compile.worker.ts:254]: We're probably going to want to clear the databases on seek, but can figure that out later.

**Review by qix — APPROVED** [review]
Awesome! I already tried it out on your branch and was looking good.

File: websqrl/src/bluesky-sample.sqrl (hunk 3 above)
Comment (meyer) [reply websqrl/src/bluesky-sample.sqrl:23]: yeh, i noticed it on `delete-like` events but there might be others as well

File: websqrl/src/EventDomain.ts (hunk 2 above)
Comment (meyer) [reply websqrl/src/EventDomain.ts:4]: good idea, can do

**Pull request merged**